File: 2024/Day04/part02.js
import fs from 'fs';

const example = `.M.S......
..A..MSMS.
.M.S.MAA..
..A.ASMSM.
.M.S.M....
..........
S.S.S.S.S.
.A.A.A.A..
M.M.M.M.M.
..........`;

const input = fs.readFileSync('input.txt', 'utf8').trim();
const grid = input.split("\n").map((row) => row.trim().split(""));
const word = "MAS".split("");

const search = (sx, sy) => (dx, dy) => {
    let { x, y, c } = { x: sx, y: sy, c: 0 };
    while (grid[x]?.[y] && word[c]) {
        if (grid[x][y] !== word[c]) break;
        x += dx, y += dy, c++;
    }
    return +(c >= word.length);
};

let count = 0, off = word.length - 1;
for (let y = 0; y < grid[0].length; y++) {
    for (let x = 0; x < grid.length; x++) {
        let a = search(x, y)(1,1) + search(x + off, y + off)(-1,-1);
        let b = search(x, y + off)(1,-1) + search(x + off, y)(-1,1);
        count += (a + b) > 1;
    }
}

console.log(count);

File: 2024/Day04/part02.rs
#![allow(unused_variables)]
use std::fs::read_to_string;

fn main () {
    let input = ".M.S......
..A..MSMS.
.M.S.MAA..
..A.ASMSM.
.M.S.M....
..........
S.S.S.S.S.
.A.A.A.A..
M.M.M.M.M.
..........";    

    let input = read_to_string("input.txt").unwrap().trim().to_string();
    let lines: Vec<&str> = input.split("\n").map(|l| l.trim()).collect();
    let grid: Vec<Vec<char>> = lines.iter().map(|l| l.chars().collect()).collect();
    let size = (grid.len() as i32, grid[0].len() as i32);
    let word = "MAS".chars().collect::<Vec<char>>();

    let search = |sx: i32, sy: i32| {
        let grid = &grid;
        let word = &word;
        move |dx: i32, dy: i32| -> i32 {
            let mut c = 0;
            let mut x = sx; 
            let mut y = sy; 
            while x >= 0 && x < size.0 && y >= 0 && y < size.1 && c < word.len() {
                if grid[x as usize][y as usize] != word[c] { break; }
                x += dx; y += dy; c += 1;
            }
            (c >= word.len()).into()
        }
    };
    
    let mut count = 0;
    let off = word.len() as i32 - 1;
    for y in 0..size.1 {
        for x in 0..size.0 {
            let a = search(x, y)(1,1) + search(x + off, y + off)(-1,-1);
            let b = search(x, y + off)(1,-1) + search(x + off, y)(-1,1);
            match a + b { 2 => count += 1, _ => () }
        }
    }

    println!("{:?}", count);
}

